askama_shared/src/generator.rs

```diff
@@ -34,7 +34,7 @@ struct Generator<'a, S: std::hash::BuildHasher> {
     // What integrations need to be generated
     integrations: Integrations,
     // Variables accessible directly from the current scope (not redirected to context)
-    locals: SetChain<'a, &'a str>,
+    locals: MapChain<'a, &'a str, Option<String>>,
```

Comment:
I feel like making `MapChain` abstract over the key type here isn't actually useful, and doing the same thing for the value argument makes it nice and general at the cost of having to deal with these nested options. Do you think we can flatten the abstraction here? It really doesn't need to be general anyway.

Reply:
So I have a related question. I've looked a bit into #329 (shadowing). My initial impression is that locals need to have an `initialized: bool`, to be able to know if it should shadow or assign. So I'm considering changing `Option<String>` into `LocalMeta`.

So for this PR, I could change it to:

```rust
struct LocalMeta {
    references: Option<String>,
}
```

Then in the PR for shadowing, I can introduce `initialized`.

```rust
struct LocalMeta {
    references: Option<String>,
    initialized: bool,
}
```

What do you think about that?

Reply:
Sounds about right. BTW, does the references value need to be `String`? Why can't it be `&str`?

Reply:
Alright, I'll change it to that.

It can't because in `write_call` it's used to store the result of `visit_attr`. Though, if `self.` doesn't need to be prefixed, then I might be able to change it to `Cow<'a, str>`.

Reply:
Let's leave it as `String` for now, probably not a big deal anyway.